cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetContext.java

```diff
@@ -28,22 +28,24 @@
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
 
 import java.util.Map;
 import java.util.Set;
 import javax.annotation.Nullable;
 
 /**
- * Implementation of {@link co.cask.cdap.api.data.DatasetContext} that allows to dynamically load datasets
+ * Implementation of {@link DatasetContext} that allows to dynamically load datasets
  * into a started {@link TransactionContext}.
  */
 public abstract class DynamicDatasetContext implements DatasetContext {
-  private static final Map<String, String> EMPTY_MAP = ImmutableMap.<String, String>of();
+  private static final Map<String, String> EMPTY_MAP = ImmutableMap.of();
   private final TransactionContext context;
   private final Set<String> allowedDatasets;
   private final DatasetFramework datasetFramework;
   private final ClassLoader classLoader;
   private final Map<String, String> runtimeArguments;
+  private final Set<String> txnInProgressDatasets = Sets.newHashSet();
```

Comment:
should initialize this in the constructor with whatever is already in the transaction context.